src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
-#include "mixer.h"
+#include <string>
+#include <unistd.h>
+
 #include "dma.h"
+#include "hardware.h"
+#include "mixer.h"
 #include "pic.h"
 #include "setup.h"
 #include "shell.h"
-#include "math.h"
-#include "regs.h"
-using namespace std;
-
-//Extra bits of precision over normal gus
-#define WAVE_FRACT 9
-#define WAVE_FRACT_MASK ((1 << WAVE_FRACT)-1)
-#define WAVE_MSWMASK ((1 << 16)-1)
-#define WAVE_LSWMASK (0xffffffff ^ WAVE_MSWMASK)
-
-//Amount of precision the volume has
-#define RAMP_FRACT (10)
-#define RAMP_FRACT_MASK ((1 << RAMP_FRACT)-1)
-
-#define GUS_BASE myGUS.portbase
-#define GUS_RATE myGUS.rate
-#define LOG_GUS 0
-
-#define VOL_SHIFT 14
-
-#define WCTRL_STOPPED			0x01
-#define WCTRL_STOP				0x02
-#define WCTRL_16BIT				0x04
-#define WCTRL_LOOP				0x08
-#define WCTRL_BIDIRECTIONAL		0x10
-#define WCTRL_IRQENABLED		0x20
-#define WCTRL_DECREASING		0x40
-#define WCTRL_IRQPENDING		0x80
-
-Bit8u adlib_commandreg;
-static MixerChannel * gus_chan;
-static Bit8u irqtable[8] = { 0, 2, 5, 3, 7, 11, 12, 15 };
-static Bit8u dmatable[8] = { 0, 1, 3, 5, 6, 7, 0, 0 };
-static Bit8u GUSRam[1024*1024]; // 1024K of GUS Ram
-static Bit16u vol16bit[4096];
-static Bit32u pantable[16];
-
-class GUSChannels;
-static void CheckVoiceIrq(void);
-
-struct GFGus {
-	Bit8u gRegSelect;
-	Bit16u gRegData;
-	Bit32u gDramAddr;
-	Bit16u gCurChannel;
-
-	Bit8u DMAControl;
-	Bit16u dmaAddr;
-	Bit8u TimerControl;
-	Bit8u SampControl;
-	Bit8u mixControl;
-	Bit8u ActiveChannels;
-	Bit32u basefreq;
-
-	struct GusTimer {
-		Bit8u value;
-		bool reached;
-		bool raiseirq;
-		bool masked;
-		bool running;
-		float delay;
-	} timers[2];
-	Bit32u rate;
-	Bitu portbase;
-	Bit8u dma1;
-	Bit8u dma2;
-
-	Bit8u irq1;
-	Bit8u irq2;
-
-	bool irqenabled;
-	bool ChangeIRQDMA;
-	// IRQ status register values
-	Bit8u IRQStatus;
-	Bit32u ActiveMask;
-	Bit8u IRQChan;
-	Bit32u RampIRQ;
-	Bit32u WaveIRQ;
-} myGUS;
-
-Bitu DEBUG_EnableDebugger(void);
-
-static void GUS_DMA_Callback(DmaChannel * chan,DMAEvent event);
-
-class GUSChannels {
+
+#define LOG_GUS 0 // set to 1 for detailed logging
+
+// Constants used in member definitions
+constexpr uint8_t ADLIB_CMD_DEFAULT = 85u;
+constexpr float AUDIO_SAMPLE_MAX = static_cast<float>(MAX_AUDIO);
+constexpr float AUDIO_SAMPLE_MIN = static_cast<float>(MIN_AUDIO);
+constexpr int BUFFER_FRAMES = 48;
+constexpr int BUFFER_SAMPLES = BUFFER_FRAMES * 2; // 2 samples/frame (left & right)
+constexpr uint32_t BYTES_PER_DMA_XFER = 8 * 1024; // 8 KB per transfer
+constexpr auto DELTA_DB = 0.002709201;             // 0.0235 dB increments
+constexpr uint8_t DMA_IRQ_ADDRESSES = 8u; // number of IRQ and DMA channels
+constexpr uint32_t ISA_BUS_THROUGHPUT = 32 * 1024 * 1024; // 32 MB/s
+constexpr uint16_t DMA_TRANSFERS_PER_S = ISA_BUS_THROUGHPUT / BYTES_PER_DMA_XFER;
+constexpr uint8_t MAX_VOICES = 32u;
+constexpr uint8_t MIN_VOICES = 14u;
+constexpr uint8_t MAX_DMA_ADDRESS = 7u;
+constexpr uint8_t MAX_IRQ_ADDRESS = 15u;
+constexpr uint8_t MIN_DMA_ADDRESS = 1u;
```

Comment:
Please confirm: MID_DMA_ADDRESS is actually 0; at least that's a value explicitly listed in .conf file.

Reply:
Ahh - indeed, it is zero.  Fixing!